playbooks/infra/deploy-vm-bastion-libvirt.yml

```diff
@@ -0,0 +1,241 @@
+#   system_rpm_link: http://example.rpm         # URL to the RPM package used for system activation
+
+# hypervisor:
+#   ansible_host: 10.1.1.1                      # IP address or hostname of the hypervisor
+
+# bastion:
+#   ansible_host: 10.1.1.1              # IP address or hostname of the hypervisor
+#   dns: 192.168.1.1                    # DNS server IP
+#   gateway: 10.1.1.254                 # Default gateway for the bastion
+#   hostname: bastion.example.com       # Hostname of the bastion
+#   net_prefix: "27"                    # Network prefix (CIDR notation)
+#   vm_bridge_name: example             # Hypervisor bridge name
+#   vm_name: bastion-vm-name            # VM name on the hypervisor
+
+# bastions:
+#   initial_user: username              # The initial user account to log into the bastion host
+#   vm_cpu: "8"                         # Number of virtual CPUs allocated to the VM
+#   vm_disk_size: 100G                  # Disk size allocated to the VM (in gigabytes)
+#   vm_external_interface: eno0         # External network interface for the VM
+#   vm_ram: "16384"                     # Amount of RAM allocated to the VM (in megabytes)
+
+# Roles/Tasks Overview:
+# 1. **Deploy VM on Hypervisor**
+#    - Download and resize VM QCOW2 images.
+#    - Apply network configurations via cloud-init.
+#    - Deploy VM using the customized image.
+#
+# 2. **Initial System Setup (Bastion Host)**
+#    - Install and activate the operating system.
+#    - Configure users, sudo access, and SSH settings.
+#
+# 3. **Setup Bastion Host**
+#    - Configure SSH keys for secure communication.
+#    - Set hostname and other critical system settings.
+#
+# Notes:
+# - This playbook assumes the hypervisor host is pre-installed and ready.
+# - Customize variables such as `vm_name`, `ansible_user`, and network settings to fit your environment.
+# - Test in a non-production environment before deploying.
+---
+- name: Deploy VM on hypervisor
+  hosts: hypervisor
+  vars:
+    location: rdu
+    vm_qcow_url: "{{ vm_qcow_rdu_url if location == 'rdu' else vm_qcow_tlv_url }}"
+    vm_name: "{{ hostvars['bastion'].vm_name }}"
+    vm_disk_name: "{{ vm_name }}_main.qcow2"
+    libvirtd_disk_path: /var/lib/libvirt/images
+    vm_net_config:
+      network:
+        version: 2
+        ethernets:
+          "{{ hostvars['bastion'].vm_external_interface }}":
+            match:
+              name: "{{ hostvars['bastion'].vm_external_interface }}"
+            addresses:
+              - "{{ hostvars['bastion'].ansible_host }}/{{ hostvars['bastion'].net_prefix }}"
+            gateway4: "{{ hostvars['bastion'].gateway }}"
+            nameservers:
+              addresses: ["{{ hostvars['bastion'].dns }}"]
+  tasks:
+
+    - name: Gather VM qcow image
+      ansible.builtin.get_url:
+        url: "{{ vm_qcow_url }}"
+        dest: "{{ libvirtd_disk_path }}/{{ vm_disk_name }}"
+        force: true
+        mode: '0640'
+
+    - name: Resize VM hard disk
```

Comment:
Why do we need to resize the disk?
Doesn't `create_vms` handle the disk creation?

Reply:
Yes, the creates_vms role creates an empty HDD, but in this case, we don't need an empty disk because I pre-pull the QCOW2 cloud image with the same name. As a result, the creates_vms role simply reuses my QCOW2 HDD image without any additional modifications